File: packages/mcp-servers/waveform-mcp/src/waveform_mcp/axi_decoder.py
"""AXI4 5-channel state machine decoder.

Channels: AW (write address), W (write data), B (write response),
          AR (read address), R (read data).

Handshake rule: transfer occurs when both VALID and READY are high
at the same rising clock edge.

Detects:
- Handshake completions (all 5 channels)
- Protocol violations (WVALID before AWVALID handshake, etc.)
- Timeout: VALID asserted for > timeout_clocks without READY
- Outstanding transaction count
"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class AXITransaction:
    txn_id: str
    channel: str
    addr: str | None
    start_time_ns: float
    end_time_ns: float | None
    status: str   # "ok" | "timeout" | "violation"
    notes: list[str] = field(default_factory=list)


@dataclass
class AXIViolation:
    time_ns: float
    channel: str
    violation_type: str
    message: str
    suggestion: str


def _get_value_at(tv: list[dict], time_ns: float) -> str | None:
    """Get the signal value at a given time (last value before or at time_ns)."""
    val = None
    for event in tv:
        if event["time"] <= time_ns:
            val = event["value"]
        else:
            break
    return val


def _find_clock_edges(clock_tv: list[dict]) -> list[float]:
    """Return timestamps of all rising clock edges."""
    edges: list[float] = []
    for i in range(1, len(clock_tv)):
        if clock_tv[i - 1]["value"] in ("0", "x") and clock_tv[i]["value"] == "1":
            edges.append(clock_tv[i]["time"])
    return edges


def decode_axi(
    signal_events: dict[str, list[dict]],  # name → compressed events
    axi_prefix: str = "",
    clock_name: str | None = None,
    timeout_cycles: int = 100,
) -> dict:
    """Decode AXI4 transactions and detect protocol violations.

    Args:
        signal_events: Compressed signal events from waveform-mcp compressor.
        axi_prefix:    Signal name prefix, e.g. "m_axi_" for m_axi_awvalid.
        clock_name:    Clock signal name for edge sampling (auto-detected if None).
        timeout_cycles: Cycles before flagging a VALID without READY as timeout.

    Returns:
        {
            "transactions":  [AXITransaction as dict],
            "violations":    [AXIViolation as dict],
            "summary":       str,
        }
    """
    pfx = axi_prefix

    def sig(name: str) -> list[dict]:
        return signal_events.get(f"{pfx}{name}", [])

    # Auto-detect clock
    if clock_name is None:
        for name in signal_events:
            if "clk" in name.lower() or "clock" in name.lower():
                clock_name = name
                break

    clock_tv = signal_events.get(clock_name or "", [])
    clock_edges = _find_clock_edges(clock_tv) if clock_tv else []

    if not clock_edges:
        return {
            "transactions": [],
            "violations":   [],
            "summary": f"No clock edges found ({clock_name or 'no clock detected'}). Cannot decode AXI transactions.",
        }

    transactions: list[dict] = []
    violations: list[dict] = []

    # ── Detect handshakes on each channel ─────────────────────────────────────
    channels = [
        ("aw", "awvalid", "awready"),
        ("w",  "wvalid",  "wready"),
        ("b",  "bvalid",  "bready"),
        ("ar", "arvalid", "arready"),
        ("r",  "rvalid",  "rready"),
    ]

    channel_handshakes: dict[str, list[float]] = {}

    for ch, valid_name, ready_name in channels:
        valid_tv = sig(valid_name)
        ready_tv = sig(ready_name)

        if not valid_tv:
            continue

        handshakes: list[float] = []
        valid_assert_time: float | None = None
        valid_cycle_count = 0

        for edge_time in clock_edges:
            valid_val = _get_value_at(valid_tv, edge_time)
            ready_val = _get_value_at(ready_tv, edge_time) if ready_tv else None

            if valid_val == "1":
                if valid_assert_time is None:
                    valid_assert_time = edge_time
                valid_cycle_count += 1

                if ready_val == "1":
                    handshakes.append(edge_time)
                    valid_assert_time = None
                    valid_cycle_count = 0
                elif valid_cycle_count > timeout_cycles:
                    violations.append({
                        "time_ns":       edge_time,
                        "channel":       ch.upper(),
                        "violation_type": "handshake_timeout",
                        "message": (
                            f"{pfx}{valid_name} asserted for >{timeout_cycles} cycles "
                            f"without {pfx}{ready_name} response"
                        ),
                        "suggestion":    "Check if slave is stalled or READY is driven correctly",
                    })
                    valid_assert_time = None
                    valid_cycle_count = 0
            else:
                valid_assert_time = None
                valid_cycle_count = 0

        channel_handshakes[ch] = handshakes

    # ── AXI ordering violation: W before AW ───────────────────────────────────
    aw_times = channel_handshakes.get("aw", [])
    w_times  = channel_handshakes.get("w", [])

    for w_time in w_times:
        preceding_aw = [t for t in aw_times if t <= w_time]
        if not preceding_aw:
            violations.append({
                "time_ns":       w_time,
                "channel":       "W",
                "violation_type": "w_before_aw",
                "message": f"Write data handshake at {w_time:.1f}ns before any write address handshake",
                "suggestion":    "Ensure AWVALID/AWREADY completes before or concurrently with WVALID/WREADY",
            })

    # ── Build transaction records (AW→W→B triples) ────────────────────────────
    b_times = channel_handshakes.get("b", [])
    for i, aw_time in enumerate(aw_times):
        next_aw = aw_times[i + 1] if i + 1 < len(aw_times) else float("inf")
        # W handshakes between this AW and the next
        matching_w = [t for t in w_times if aw_time <= t < next_aw]
        # B handshake after AW
        matching_b = [t for t in b_times if t > aw_time]
        b_time = matching_b[0] if matching_b else None

        addr_sig = sig("awaddr")
        addr_val = _get_value_at(addr_sig, aw_time) if addr_sig else None

        transactions.append({
            "txn_id":     f"write_{i}",
            "type":       "write",
            "aw_time_ns": aw_time,
            "w_times_ns": matching_w,
            "b_time_ns":  b_time,
            "addr":       addr_val,
            "status":     "ok" if b_time else "no_response",
        })

    ar_times = channel_handshakes.get("ar", [])
    r_times  = channel_handshakes.get("r", [])
    for i, ar_time in enumerate(ar_times):
        next_ar = ar_times[i + 1] if i + 1 < len(ar_times) else float("inf")
        matching_r = [t for t in r_times if ar_time < t < next_ar]

        addr_sig = sig("araddr")
        addr_val = _get_value_at(addr_sig, ar_time) if addr_sig else None

        transactions.append({
            "txn_id":     f"read_{i}",
            "type":       "read",
            "ar_time_ns": ar_time,
            "r_times_ns": matching_r,
            "addr":       addr_val,
            "status":     "ok" if matching_r else "no_response",
        })

    # ── Summary ───────────────────────────────────────────────────────────────
    writes = sum(1 for t in transactions if t["type"] == "write")
    reads  = sum(1 for t in transactions if t["type"] == "read")
    summary = (
        f"AXI decode: {writes} write transactions, {reads} read transactions, "
        f"{len(violations)} protocol violation(s) on prefix '{pfx}'"
    )

    return {
        "transactions": transactions,
        "violations":   violations,
        "summary":      summary,
    }

File: packages/mcp-servers/waveform-mcp-rs/src/axi_decoder.rs
//! AXI4 5-channel state machine decoder.
//!
//! Port of `axi_decoder.py` from the deleted Python `waveform-mcp` package.
//! Reads compressed signal events for an AXI bus, samples valid/ready at each
//! rising clock edge, and emits both completed transactions and protocol
//! violations.
//!
//! Channels: AW (write addr), W (write data), B (write response),
//!           AR (read addr), R (read data).

use serde_json::{json, Map, Value};

const AXI_CHANNELS: &[(&str, &str, &str)] = &[
    ("aw", "awvalid", "awready"),
    ("w", "wvalid", "wready"),
    ("b", "bvalid", "bready"),
    ("ar", "arvalid", "arready"),
    ("r", "rvalid", "rready"),
];

fn event_time(ev: &Value) -> f64 {
    ev.get("time").and_then(|v| v.as_f64()).unwrap_or(0.0)
}

fn event_value(ev: &Value) -> &str {
    ev.get("value").and_then(|v| v.as_str()).unwrap_or("")
}

/// Last value at or before `t_ns`.
fn value_at(events: &[Value], t_ns: f64) -> Option<&str> {
    let mut val: Option<&str> = None;
    for ev in events {
        if event_time(ev) <= t_ns {
            val = Some(event_value(ev));
        } else {
            break;
        }
    }
    val
}

fn rising_edges(events: &[Value]) -> Vec<f64> {
    let mut edges = Vec::new();
    for i in 1..events.len() {
        let prev = event_value(&events[i - 1]);
        let cur = event_value(&events[i]);
        if (prev == "0" || prev == "x") && cur == "1" {
            edges.push(event_time(&events[i]));
        }
    }
    edges
}

pub fn decode_axi(
    signal_events: &Map<String, Value>,
    axi_prefix: &str,
    clock_name: Option<&str>,
    timeout_cycles: u32,
) -> Value {
    let sig = |suffix: &str| -> Vec<Value> {
        signal_events
            .get(&format!("{}{}", axi_prefix, suffix))
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default()
    };

    // Auto-detect clock if not provided.
    let detected_clock: Option<String> = clock_name.map(|s| s.to_string()).or_else(|| {
        signal_events.keys().find_map(|n| {
            let lc = n.to_lowercase();
            if lc.contains("clk") || lc.contains("clock") {
                Some(n.clone())
            } else {
                None
            }
        })
    });

    let clock_events = detected_clock
        .as_ref()
        .and_then(|c| signal_events.get(c))
        .and_then(|v| v.as_array())
        .cloned()
        .unwrap_or_default();

    let edges = rising_edges(&clock_events);
    if edges.is_empty() {
        return json!({
            "transactions": [],
            "violations":   [],
            "summary":      format!(
                "No clock edges found ({}). Cannot decode AXI transactions.",
                detected_clock.unwrap_or_else(|| "no clock detected".to_string())
            ),
        });
    }

    let mut transactions: Vec<Value> = Vec::new();
    let mut violations: Vec<Value> = Vec::new();
    let mut channel_handshakes: std::collections::HashMap<&str, Vec<f64>> =
        std::collections::HashMap::new();

    for (ch, valid_name, ready_name) in AXI_CHANNELS {
        let valid_tv = sig(valid_name);
        if valid_tv.is_empty() {
            continue;
        }
        let ready_tv = sig(ready_name);

        let mut handshakes: Vec<f64> = Vec::new();
        let mut valid_cycle_count: u32 = 0;

        for &edge in &edges {
            let valid_val = value_at(&valid_tv, edge).unwrap_or("0");
            let ready_val = if ready_tv.is_empty() {
                None
            } else {
                value_at(&ready_tv, edge)
            };

            if valid_val == "1" {
                valid_cycle_count += 1;
                if ready_val == Some("1") {
                    handshakes.push(edge);
                    valid_cycle_count = 0;
                } else if valid_cycle_count > timeout_cycles {
                    violations.push(json!({
                        "time_ns":         edge,
                        "channel":         ch.to_uppercase(),
                        "violation_type":  "handshake_timeout",
                        "message": format!(
                            "{}{} asserted for >{} cycles without {}{} response",
                            axi_prefix, valid_name, timeout_cycles, axi_prefix, ready_name
                        ),
                        "suggestion": "Check if slave is stalled or READY is driven correctly",
                    }));
                    valid_cycle_count = 0;
                }
            } else {
                valid_cycle_count = 0;
            }
        }

        channel_handshakes.insert(*ch, handshakes);
    }

    // Ordering violation: W before AW.
    let empty_vec: Vec<f64> = Vec::new();
    let aw_times = channel_handshakes.get("aw").unwrap_or(&empty_vec).clone();
    let w_times = channel_handshakes.get("w").unwrap_or(&empty_vec).clone();
    let b_times = channel_handshakes.get("b").unwrap_or(&empty_vec).clone();
    let ar_times = channel_handshakes.get("ar").unwrap_or(&empty_vec).clone();
    let r_times = channel_handshakes.get("r").unwrap_or(&empty_vec).clone();

    for &w_t in &w_times {
        if !aw_times.iter().any(|&aw| aw <= w_t) {
            violations.push(json!({
                "time_ns":         w_t,
                "channel":         "W",
                "violation_type":  "w_before_aw",
                "message": format!(
                    "Write data handshake at {:.1}ns before any write address handshake",
                    w_t
                ),
                "suggestion": "Ensure AWVALID/AWREADY completes before or concurrently with WVALID/WREADY",
            }));
        }
    }

    // Build write transactions (AW → matching W*, B?).
    let awaddr_tv = sig("awaddr");
    for (i, &aw_t) in aw_times.iter().enumerate() {
        let next_aw = aw_times.get(i + 1).copied().unwrap_or(f64::INFINITY);
        let matching_w: Vec<f64> = w_times.iter().copied().filter(|t| *t >= aw_t && *t < next_aw).collect();
        let b_t = b_times.iter().copied().find(|t| *t > aw_t);
        let addr = if awaddr_tv.is_empty() {
            Value::Null
        } else {
            value_at(&awaddr_tv, aw_t)
                .map(|s| Value::String(s.to_string()))
                .unwrap_or(Value::Null)
        };
        transactions.push(json!({
            "txn_id":     format!("write_{}", i),
            "type":       "write",
            "aw_time_ns": aw_t,
            "w_times_ns": matching_w,
            "b_time_ns":  b_t,
            "addr":       addr,
            "status":     if b_t.is_some() { "ok" } else { "no_response" },
        }));
    }

    // Build read transactions (AR → matching R*).
    let araddr_tv = sig("araddr");
    for (i, &ar_t) in ar_times.iter().enumerate() {
        let next_ar = ar_times.get(i + 1).copied().unwrap_or(f64::INFINITY);
        let matching_r: Vec<f64> = r_times.iter().copied().filter(|t| *t > ar_t && *t < next_ar).collect();
        let addr = if araddr_tv.is_empty() {
            Value::Null
        } else {
            value_at(&araddr_tv, ar_t)
                .map(|s| Value::String(s.to_string()))
                .unwrap_or(Value::Null)
        };
        transactions.push(json!({
            "txn_id":     format!("read_{}", i),
            "type":       "read",
            "ar_time_ns": ar_t,
            "r_times_ns": matching_r,
            "addr":       addr,
            "status":     if matching_r.is_empty() { "no_response" } else { "ok" },
        }));
    }

    let writes = transactions.iter().filter(|t| t["type"] == "write").count();
    let reads = transactions.iter().filter(|t| t["type"] == "read").count();
    let summary = format!(
        "AXI decode: {} write transactions, {} read transactions, {} protocol violation(s) on prefix '{}'",
        writes,
        reads,
        violations.len(),
        axi_prefix
    );

    json!({
        "transactions": transactions,
        "violations":   violations,
        "summary":      summary,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn evt(t: f64, v: &str) -> Value {
        json!({"time": t, "value": v})
    }

    fn make_axi(events: Vec<(&str, Vec<(f64, &str)>)>) -> Map<String, Value> {
        let mut m = Map::new();
        for (name, points) in events {
            m.insert(
                name.to_string(),
                Value::Array(points.into_iter().map(|(t, v)| evt(t, v)).collect()),
            );
        }
        m
    }

    #[test]
    fn decode_single_write_transaction() {
        // Single AW + W + B handshake at distinct clock edges.
        let events = make_axi(vec![
            (
                "aclk",
                vec![
                    (0.0, "0"),
                    (5.0, "1"),
                    (10.0, "0"),
                    (15.0, "1"),
                    (20.0, "0"),
                    (25.0, "1"),
                    (30.0, "0"),
                    (35.0, "1"),
                ],
            ),
            ("m_axi_awvalid", vec![(0.0, "0"), (5.0, "1"), (6.0, "0")]),
            ("m_axi_awready", vec![(0.0, "1")]),
            ("m_axi_awaddr", vec![(0.0, "00"), (5.0, "42")]),
            ("m_axi_wvalid", vec![(0.0, "0"), (15.0, "1"), (16.0, "0")]),
            ("m_axi_wready", vec![(0.0, "1")]),
            ("m_axi_bvalid", vec![(0.0, "0"), (25.0, "1"), (26.0, "0")]),
            ("m_axi_bready", vec![(0.0, "1")]),
        ]);
        let result = decode_axi(&events, "m_axi_", Some("aclk"), 100);
        let txns = result["transactions"].as_array().unwrap();
        assert_eq!(txns.len(), 1);
        assert_eq!(txns[0]["type"], "write");
        assert_eq!(txns[0]["status"], "ok");
        assert_eq!(result["violations"].as_array().unwrap().len(), 0);
    }

    #[test]
    fn detect_handshake_timeout() {
        // VALID stays high forever, READY never asserts → timeout violation.
        let mut clk_events = Vec::new();
        for i in 0..50 {
            let t = i as f64 * 5.0;
            clk_events.push((t, if i % 2 == 0 { "0" } else { "1" }));
        }
        let events = make_axi(vec![
            ("aclk", clk_events),
            ("m_axi_awvalid", vec![(0.0, "1")]),
            ("m_axi_awready", vec![(0.0, "0")]),
        ]);
        let result = decode_axi(&events, "m_axi_", Some("aclk"), 5);
        let violations = result["violations"].as_array().unwrap();
        assert!(violations.len() >= 1);
        assert_eq!(violations[0]["violation_type"], "handshake_timeout");
    }
}
